src/Emulator/Peripherals/Peripherals/MTD/OpenTitan_FlashController.cs

```diff
@@ -0,0 +1,1185 @@
+            ** It currently doesnt have a unit test though.
+            ** BANK*_INFO*_PAGE_CFG_* registers are of interest
+            */
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoProgramInfo: busWordAddress = 0x{0:X}",
+                busWordAddress);
+            uint value = programFifo.Dequeue();
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoProgramInfo: value = 0x{0:X}",
+                value);
+            infoFlash.WriteDoubleWord(busWordAddress - FLASH_MEM_BASE_ADDR, value);
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoProgramInfo: programFifo.Count = {0}",
+                programFifo.Count);
+        }
+
+        private bool IsOperationAllowedInDefaultRegion(OperationType opType){
+            bool ret = false;
+            uint DEFAULT_REGION_VALUE = registersCollection.Read((long)Registers.DEFAULT_REGION);
+            uint DEFAULT_REGION_READ_EN = 0x1 & DEFAULT_REGION_VALUE;
+            uint DEFAULT_REGION_PROG_EN = (0x1 << 1) & DEFAULT_REGION_VALUE;
+            uint DEFAULT_REGION_ERASE_EN = (0x1 << 2) & DEFAULT_REGION_VALUE;
+
+            if(opType == OperationType.READ_DATA && DEFAULT_REGION_READ_EN != 0){
+                ret = true;
+            }else if(opType == OperationType.PROGRAM_DATA && DEFAULT_REGION_PROG_EN != 0){
+                ret = true;
+            }else if(opType == OperationType.ERASE_DATA_PAGE && DEFAULT_REGION_ERASE_EN != 0){
+                ret = true;
+            }
+
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/IsOperationAllowedInDefaultRegion: ret = {0}",
+                ret);
+
+            if(!ret){
+                this.Log(
+                LogLevel.Info,
+                    "OpenTitan_FlashController/IsOperationAllowedInDefaultRegion: " +
+                    "Operation not allowed in the default region");
+            }
+
+            return ret;
+        }
+
+        private bool IsOperationAllowed(OperationType opType, long operationAddress){
+            bool ret = IsOperationAllowedInDefaultRegion(opType);
+
+            long[] protectionRegionRegisters = new long[] {
+                (long)Registers.MP_REGION_CFG_0,
+                (long)Registers.MP_REGION_CFG_1,
+                (long)Registers.MP_REGION_CFG_2,
+                (long)Registers.MP_REGION_CFG_3,
+                (long)Registers.MP_REGION_CFG_4,
+                (long)Registers.MP_REGION_CFG_5,
+                (long)Registers.MP_REGION_CFG_6,
+                (long)Registers.MP_REGION_CFG_7
+            };
+
+            foreach(long mpRegisterOffset in protectionRegionRegisters){
+                if(MpRegionRegisterAppliesToOperation(mpRegisterOffset, operationAddress)){
+                    ret = IsOperationAllowedInMpRegion(opType, mpRegisterOffset);
```

Comment:
Is it possible that `ret` is set here many times?

Should the loop be broken at first hit or should the result be overwritten/ored/anded?

Reply:
Good question. Ill see if I can figure out what the RTL does if multiple MP registers for the same address disagree.

If i can't figure out the RTL I will change it to break at first hit.

Reply:
https://github.com/lowRISC/opentitan/blob/d9dc048680d89bd630ffae673103509e28017709/hw/ip/flash_ctrl/rtl/flash_mp_data_region_sel.sv#L20

Judging by the comment it is okay to break the loop as that will prioritize the lowest index, similar to how the RTL does